kiwi/lib/search.py

```diff
@@ -0,0 +1,438 @@
+        names.append('hter')
+        values.append(hter)
+
+    if config.options.search_word_level:
+        assert kiwi_config['data']['train']['output']['target_tags'] is not None
+        assert kiwi_config['data']['valid']['output']['target_tags'] is not None
+        word_level = trial.suggest_categorical('word_level', [True, False])
+        kiwi_config['system']['model']['outputs']['word_level']['target'] = word_level
+        kiwi_config['system']['model']['outputs']['word_level']['gaps'] = word_level
+        names.append('word_level')
+        values.append(word_level)
+
+    if config.options.search_hter and config.options.search_word_level:
+        if hter and word_level and config.options.search_sentence_loss_weight:
+            # Also search for the sentence weight
+            sentence_loss_weight = trial.suggest_uniform(
+                'sentence_loss_weight',
+                config.options.sentence_loss_weight.lower,
+                config.options.sentence_loss_weight.upper,
+            )
+            kiwi_config['system']['model']['outputs'][
+                'sentence_loss_weight'
+            ] = sentence_loss_weight
+            names.append('sentence_loss_weight')
+            values.append(sentence_loss_weight)
+
+    specified_word_level = kiwi_config['system']['model']['outputs']['word_level'][
+        'target'
+    ]
+    if specified_word_level and config.options.search_hter:
+        if hter and config.options.search_sentence_loss_weight:
+            # Also search for the sentence weight
+            sentence_loss_weight = trial.suggest_uniform(
+                'sentence_loss_weight',
+                config.options.sentence_loss_weight.lower,
+                config.options.sentence_loss_weight.upper,
+            )
+            kiwi_config['system']['model']['outputs'][
+                'sentence_loss_weight'
+            ] = sentence_loss_weight
+            names.append('sentence_loss_weight')
+            values.append(sentence_loss_weight)
+
+    specified_sentence_level = kiwi_config["system"]["model"]["outputs"][
+        "sentence_level"
+    ]["hter"]
+    if specified_sentence_level and config.options.search_sentence_loss_weight:
+        # Also search for the sentence weight
+        sentence_loss_weight = trial.suggest_uniform(
+            'sentence_loss_weight',
+            config.options.sentence_loss_weight.lower,
+            config.options.sentence_loss_weight.upper,
+        )
+        kiwi_config["system"]["model"]["outputs"][
+            "sentence_loss_weight"
+        ] = sentence_loss_weight
+        names.append("sentence_loss_weight")
+        values.append(sentence_loss_weight)
+
+    if kiwi_config['quiet'] is None:
+        kiwi_config['quiet'] = False
+    if kiwi_config['verbose'] is None:
+        kiwi_config['verbose'] = False
+    trainconfig = train.Configuration(**kiwi_config)
+
+    # TODO: integrate PTL trainer callback
```

Comment:
FYI: not really a comment, but I have integrated this callback and it works much worse than early stopping, hence why it didn't make the kiwi2 release :/

Reply:
Ok, so let's keep early stopping as the method of killing search runs!